float32 rota
float32 upper
